File: proposed-projects/round-trip-experiment/reverse/20260909T123239Z_5982b634/20260910T070936Z_fb630842/ported_implementation/gbnf/grammar_graph/get_serialized_rule_key.py
from __future__ import annotations

import json

from .type_guards import is_rule_char, is_rule_char_exclude, is_rule_end, is_rule_ref
from .types import RuleType

KEY_TRANSLATION = {
    RuleType.END: 0,
    RuleType.CHAR: 1,
    RuleType.CHAR_EXCLUDE: 2,
    RuleType.REF: 3,
}


def _stringify(value) -> str:
    return json.dumps(value, separators=(',', ':'))


def get_serialized_rule_key(rule) -> str:
    if is_rule_end(rule):
        return f'{KEY_TRANSLATION[RuleType.END]}'

    if is_rule_char(rule):
        return f'{KEY_TRANSLATION[RuleType.CHAR]}-{_stringify(rule.value)}'

    if is_rule_char_exclude(rule):
        return f'{KEY_TRANSLATION[RuleType.CHAR_EXCLUDE]}-{_stringify(rule.value)}'

    if is_rule_ref(rule):
        return f'{KEY_TRANSLATION[RuleType.REF]}-{rule.value}'

    raise ValueError(f'Unknown rule type: {rule!r}')

File: packages/gbnf-experiment/data/20260909T123239Z_5982b634/ported_implementation/src/grammar-graph/get-serialized-rule-key.ts
import { isRuleChar, isRuleCharExclude, isRuleEnd, isRuleRef } from './type-guards.js';
import { RuleType, type UnresolvedRule } from './types.js';

const KEY_TRANSLATION = {
  [RuleType.END]: 0,
  [RuleType.CHAR]: 1,
  [RuleType.CHAR_EXCLUDE]: 2,
  [RuleType.REF]: 3,
} as const;

export const getSerializedRuleKey = (rule: UnresolvedRule): string => {
  if (isRuleEnd(rule)) {
    return `${KEY_TRANSLATION[RuleType.END]}`;
  }

  if (isRuleChar(rule)) {
    return `${KEY_TRANSLATION[RuleType.CHAR]}-${JSON.stringify(rule.value)}`;
  }

  if (isRuleCharExclude(rule)) {
    return `${KEY_TRANSLATION[RuleType.CHAR_EXCLUDE]}-${JSON.stringify(rule.value)}`;
  }

  if (isRuleRef(rule)) {
    return `${KEY_TRANSLATION[RuleType.REF]}-${rule.value}`;
  }

  throw new Error(`Unknown rule type: ${JSON.stringify(rule)}`);
};
